compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -368,11 +391,103 @@ impl<'tcx> SizeSkeleton<'tcx> {
             (SizeSkeleton::Pointer { tail: a, .. }, SizeSkeleton::Pointer { tail: b, .. }) => {
                 a == b
             }
+            // constants are always pre-normalized into a canonical form so this
+            // only needs to check if their pointers are identical.
+            (SizeSkeleton::Generic(a), SizeSkeleton::Generic(b)) => a == b,
             _ => false,
         }
     }
 }
 
+fn combine_sorted_expr<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+    a: ty::Const<'tcx>,
+    b: ty::Const<'tcx>,
+) -> ty::Const<'tcx> {
+    use crate::mir::BinOp::Mul;
+    use ty::ConstKind::Expr;
+    use ty::Expr::Binop;
+
+    // fast-path
+    match (a.kind(), b.kind()) {
+        (Expr(Binop(Mul, ..)), _) | (_, Expr(Binop(Mul, ..))) => {}
+        _ => {
+            let (l, r) = if a < b { (a, b) } else { (b, a) };
+            return tcx.mk_const(Expr(Binop(Mul, l, r)), tcx.types.usize);
+        }
+    }
+
+    let mut buf = vec![];
```

Comment:
You probably want to use a `SmallVec`, given all the trouble you've went through to avoid calling `.sort()` (for performance, I assume?).